### PR TITLE
add sana-sprint

### PR DESCRIPTION
**Pull request by yiyixuxu**
```python
# test sana sprint
"""
python scripts/convert_sana_to_diffusers.py \
  --orig_ckpt_path /raid/.cache/huggingface/yiyi/models--JunsongChen--Sana_Sprint_1600M_1024px/snapshots/8ecfdc7e6269e5b065f5b2cf3fac9a2a1a778c6a/Sana_Sprint_1600M_1024px_36K.pth \
  --model_type SanaSprint_1600M_P1_D20 \
  --image_size 1024 \
  --dump_path /raid/yiyi/Sana-Sprint-yiyi \
  --save_full_pipeline \
  --scheduler_type scm
"""

from diffusers import SanaSprintPipeline
import torch

device = "cuda:0"
dtype = torch.bfloat16

repo = "/raid/yiyi/Sana-Sprint-yiyi"

pipeline = SanaSprintPipeline.from_pretrained(repo, torch_dtype=dtype)
pipeline.to(device)


prompt = "a tiny astronaut hatching from an egg on the moon"

image = pipeline(prompt=prompt, num_inference_steps=2).images[0]
image.save("test_out.png")
```

yiyi_test_8_out

vibe tests with different timesteps settings 

```python
# test sana sprint
# (pipeline)
test_max_timesteps = [1.57080, 1.56830, 1.56580, 1.56454, 1.56246, 1.55830, 1.55413, 1.55080, 1.54580]
test_intermediate_timesteps = [None, 1.0, 1.1, 1.2, 1.3, 1.4]
test_num_inference_steps = [1,2,4]

# test_max_timesteps = [1.57080]
# test_intermediate_timesteps = [None]
# test_num_inference_steps = [1]

from diffusers import SanaSprintPipeline
import torch

device = "cuda:0"
dtype = torch.bfloat16
repo = "/raid/yiyi/Sana-Sprint-yiyi"

def run_sana(pipeline, num_inference_steps, max_timesteps, intermediate_timesteps):
    prompt = "a tiny astronaut hatching from an egg on the moon"
    generator = torch.Generator(device=device).manual_seed(123)
    test_name = f"num_inference_steps_{num_inference_steps}_max_timesteps_{max_timesteps}_intermediate_timesteps_{intermediate_timesteps}"
    print(f"--------------------------------")
    print(f"Running test:")
    print(f"num_inference_steps: {num_inference_steps}")
    print(f"max_timesteps: {max_timesteps}")
    print(f"intermediate_timesteps: {intermediate_timesteps}")
    try:
        image = pipeline(prompt=prompt, num_inference_steps=num_inference_steps, max_timesteps=max_timesteps, intermediate_timesteps=intermediate_timesteps, generator=generator).images[0]
        image.save(f"yiyi_test_10_1_out_{test_name}.png")
    except Exception as e:
        print(e)
    print(f"--------------------------------")


pipeline = SanaSprintPipeline.from_pretrained(repo, torch_dtype=dtype)
pipeline.to(device)

for num_inference_steps in test_num_inference_steps:
    for max_timesteps in test_max_timesteps:
        for intermediate_timesteps in test_intermediate_timesteps:
            run_sana(pipeline, num_inference_steps, max_timesteps, intermediate_timesteps)
```

### CODE REVIEW

**Comment by ishan-modi** [comment]
Nice Work !! just a heads up this PR might have conflicts with #11040 if merged first

**Comment by lawrence-cj** [comment]
Since SANA-Sprint and SANA-1.5 follow the same architecture, so this PR would make SANA-1.5 work as well. 
@yiyixuxu

**Comment by yiyixuxu** [comment]
@bot /style

**Comment by yiyixuxu** [comment]
@bot/ style

File: src/diffusers/models/transformers/sana_transformer.py [hunk 1]
```diff
@@ -96,6 +96,101 @@ def forward(
         return hidden_states
 
 
+class SanaCombinedTimestepGuidanceEmbeddings(nn.Module):
+    """
+    For Sana.
+
+    Reference:
+    """
+
```
Comment (a-r-r-o-w) [inline src/diffusers/models/transformers/sana_transformer.py:105]:
```suggestion
```

No need to document extra information as these are already done in the main transformer docstring

File: src/diffusers/schedulers/scheduling_scm.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+
+    Args:
+        num_train_timesteps (`int`, defaults to 1000):
+            The number of diffusion steps to train the model.
+        prediction_type (`str`, defaults to `trigflow`):
+            Prediction type of the scheduler function. Currently only supports "trigflow".
+        max_timesteps (`float`, defaults to 1.57080):
+            The maximum timestep value used in the diffusion process.
+        intermediate_timesteps (`float`, *optional*, defaults to 1.3):
+            The intermediate timestep value used when num_inference_steps=2.
```
Comment (a-r-r-o-w) [inline src/diffusers/schedulers/scheduling_scm.py:65]:
```suggestion
```

File: src/diffusers/schedulers/scheduling_scm.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+            model_output (`torch.FloatTensor`):
+                The direct output from learned diffusion model.
+            timestep (`float`):
+                The current discrete timestep in the diffusion chain.
+            sample (`torch.FloatTensor`):
+                A current instance of a sample created by the diffusion process.
+            return_dict (`bool`, *optional*, defaults to `True`):
+                itself. Useful for methods such as [`CycleDiffusion`].
+            return_dict (`bool`, *optional*, defaults to `True`):
+                Whether or not to return a [`~schedulers.scheduling_lcm.LCMSchedulerOutput`] or `tuple`.
```
Comment (a-r-r-o-w) [inline src/diffusers/schedulers/scheduling_scm.py:225]:
```suggestion
                Whether or not to return a [`~schedulers.scheduling_scm.SCMSchedulerOutput`] or `tuple`.
```

File: src/diffusers/schedulers/scheduling_scm.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+        Args:
+            model_output (`torch.FloatTensor`):
+                The direct output from learned diffusion model.
+            timestep (`float`):
+                The current discrete timestep in the diffusion chain.
+            sample (`torch.FloatTensor`):
+                A current instance of a sample created by the diffusion process.
+            return_dict (`bool`, *optional*, defaults to `True`):
+                itself. Useful for methods such as [`CycleDiffusion`].
```
Comment (a-r-r-o-w) [inline src/diffusers/schedulers/scheduling_scm.py:223]:
```suggestion
```

File: src/diffusers/pipelines/sana/pipeline_sana_sprint.py [hunk 5]
```diff
@@ -0,0 +1,942 @@
...
+    "4.0": [8192.0, 2048.0],
+}
+
+EXAMPLE_DOC_STRING = """
+    Examples:
+        ```py
+        >>> import torch
+        >>> from diffusers import SanaPipeline
+
+        >>> pipe = SanaPipeline.from_pretrained(
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/sana/pipeline_sana_sprint.py:116]: Example to be updated to SanaSprintPipeline

File: src/diffusers/pipelines/sana/pipeline_sana_sprint.py [hunk 6]
```diff
@@ -0,0 +1,942 @@
...
+        accepts_eta = "eta" in set(inspect.signature(self.scheduler.step).parameters.keys())
+        extra_step_kwargs = {}
+        if accepts_eta:
+            extra_step_kwargs["eta"] = eta
+
+        # check if the scheduler accepts generator
+        accepts_generator = "generator" in set(inspect.signature(self.scheduler.step).parameters.keys())
+        if accepts_generator:
+            extra_step_kwargs["generator"] = generator
+        return extra_step_kwargs
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/sana/pipeline_sana_sprint.py:413]:
```suggestion
```

Let's remove since eta is not supported, no?

File: src/diffusers/pipelines/sana/pipeline_sana_sprint.py [hunk 7]
```diff
@@ -0,0 +1,942 @@
...
+        )
+
+        latents = latents * self.scheduler.config.sigma_data
+
+        guidance = torch.full([1], guidance_scale, device=device, dtype=torch.float32)
+        guidance = guidance.expand(latents.shape[0]).to(prompt_embeds.dtype)
+        guidance = guidance * self.transformer.config.guidance_embeds_scale
+
+        # 6. Prepare extra step kwargs. TODO: Logic should ideally just be moved out of the pipeline
+        extra_step_kwargs = self.prepare_extra_step_kwargs(generator, eta)
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/sana/pipeline_sana_sprint.py:857]:
```suggestion
```

**Review by a-r-r-o-w — APPROVED** [review]
Really amazing work! Can't wait for the release :heart:

File: src/diffusers/pipelines/sana/pipeline_sana_sprint.py [hunk 8]
```diff
@@ -0,0 +1,942 @@
...
+                    xm.mark_step()
+
+        latents = denoised / self.scheduler.config.sigma_data
+        if output_type == "latent":
+            image = latents
+        else:
+            latents = latents.to(self.vae.dtype)
+            try:
+                image = self.vae.decode(latents / self.vae.config.scaling_factor, return_dict=False)[0]
+            except torch.cuda.OutOfMemoryError as e:
```
Comment (hlky) [inline src/diffusers/pipelines/sana/pipeline_sana_sprint.py:924]: Nice!

For XPU we need to use `torch.OutOfMemoryError`, also looks like that will work on CUDA.

https://github.com/pytorch/pytorch/blob/00a2c68f67adbd38847845016fd1ab9275cefbab/test/test_xpu.py#L446
https://github.com/pytorch/pytorch/blob/00a2c68f67adbd38847845016fd1ab9275cefbab/test/test_cuda.py#L3950
https://github.com/pytorch/pytorch/blob/00a2c68f67adbd38847845016fd1ab9275cefbab/test/test_cuda.py#L4154

**Review by hlky — APPROVED** [review]
Thanks @yiyixuxu

File: scripts/convert_sana_to_diffusers.py [hunk 9]
```diff
@@ -96,14 +126,22 @@ def main(args):
         flow_shift = 3.0
 
     # model config
-    if args.model_type == "SanaMS_1600M_P1_D20":
+    if args.model_type == "SanaMS_1600M_P1_D20" or args.model_type == "SanaSprint_1600M_P1_D20":
```
Comment (yiyixuxu) [inline scripts/convert_sana_to_diffusers.py:129]:
```suggestion
    if args.model_type in ["SanaMS_1600M_P1_D20", "SanaSprint_1600M_P1_D20", "SanaMS1.5_1600M_P1_D20"]
```

File: scripts/convert_sana_to_diffusers.py [hunk 10]
```diff
@@ -96,14 +126,22 @@ def main(args):
...
         layer_num = 20
     elif args.model_type == "SanaMS_600M_P1_D28":
```
Comment (yiyixuxu) [inline scripts/convert_sana_to_diffusers.py:131]:
```suggestion
    elif args.model_type in ["SanaMS_600M_P1_D28", "SanaSprint_600M_P1_D28"]:
```

**Comment by lawrence-cj** [comment]
Comments here: https://github.com/huggingface/diffusers/pull/11131
Official model: https://huggingface.co/Efficient-Large-Model/Sana_Sprint_1.6B_1024px_diffusers

File: src/diffusers/pipelines/sana/pipeline_sana_sprint.py (hunk 8 above)
Comment (sayakpaul) [reply src/diffusers/pipelines/sana/pipeline_sana_sprint.py:924]: +1 on this.

File: src/diffusers/pipelines/sana/pipeline_sana_sprint.py [hunk 11]
```diff
@@ -0,0 +1,942 @@
...
+                warnings.warn(
```
Comment (sayakpaul) [inline src/diffusers/pipelines/sana/pipeline_sana_sprint.py:925]: Should this be `logger.warning()`?

File: src/diffusers/schedulers/scheduling_scm.py [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+            elif isinstance(timesteps, torch.Tensor):
+                self.timesteps = timesteps.to(device).float()
+            else:
+                raise ValueError(f"Unsupported timesteps type: {type(timesteps)}")
+        elif intermediate_timesteps is not None:
+            self.timesteps = torch.tensor([max_timesteps, intermediate_timesteps, 0], device=device).float()
+        else:
+            # max_timesteps=arctan(80/0.5)=1.56454 is the default from sCM paper, we choose a different value here
+            self.timesteps = torch.linspace(max_timesteps, 0, num_inference_steps + 1, device=device).float()
+        print(f"Set timesteps: {self.timesteps}")
```
Comment (sayakpaul) [inline src/diffusers/schedulers/scheduling_scm.py:171]:
```suggestion
```
